### ISSUE TITLE
Poor performance (maybe caused be inheritance)

### ISSUE THREAD

**Issue by georg-tiefenbrunn** [FluidTYPO3/fluidpages#255]
Hi FluidTYPO3-Gurus!

I updated all FluidTYPO3 extensions on my development VM (git dev on 6.2-dev) and noticed a huge loss in performance. My provider extension should be ready for Flux 7.2 (ViewContext and TemplatePaths). The expected content gets renderd, but the page generation time of ``page.5`` is badly high.

I first removed all viewHelpers, then tried with newly created pages without any content elements, then  disabled rendering of all partials in ``Layouts/Page.html`` to exclude EXT:fluidcontent as a cause. Working with pages in the BE I also noticed reduced performance when editing page properties which hints to a problem with EXT:flux/EXT:fluidpages. So I removed all xml data from ``tx_fed_page_flexform`` and ``tx_fed_page_flexform_sub`` fields to be sure no data or structure error causes the performance loss. My last "fast working" revision was 3648871555be7cd49b87560eae841f5d5548078d. After looking at the changes in the lat two month I think the loss of performance is caused by inheritance.

Of course all caches (file, DB) were cleared multiple times. Strangely enough some pages seemed not affected. Using the admin panel I recorded some page generation times of one page C

1. first in the root level of the website (3 levels under page id=0) 
2. page C in page B (C is 4 levels under page id=0) and 
3. page C in page B in page A (C is 5 levels under page id=0)

The more parent pages the higher was the performance loss:

My last "fast working" revision 3648871555be7cd49b87560eae841f5d5548078d
~~~~
TypoScript Key		Value	Time	Own		Sub		Total
Page generation				96		32		+433	=465
	page			PAGE	128		0		+433	=433
		5			USER	128		433
~~~~

Using latest git-dev versions of FluidTYPO3 

3 levels under page id=0
~~~~
TypoScript Key		Value	Time	Own		Sub		Total
Page generation				52		0		+4649	=4649
	page			PAGE	52		1		+4648	=4649
		5			USER	52		4648
~~~~

4 levels under page id=0
~~~~
TypoScript Key		Value	Time	Own		Sub		Total
Page generation				77		4		+112778	=112782
	page			PAGE	78		1		+112777	=112778
		5			USER	78		112777
~~~~

I stopped to track 5 levels under page id=0 after five minutes.

Any hints what I could/should do?

**Comment by InvisibleKind** [comment]
I confirm this issue. Struggling with same behaviour since latest Flux-family updates. Working with Xdebug I found following behaviour - lets say, you have 4 levels of inheritance:
```
Home
|- PageL1
  |- PageL2
    |- PageL3
      |- PageL4
```

And you have such settings in your templates:
settings.footerPid
settings.headerPid
settings.menuPid
settings.whateverPid

Now FE starts rendering of PageL4 and wants to get settings of this page.
At some point it calls PageProvider->getInheritedPropertyValueByDottedPath() for a first setting (settings.footerPid).
It calls then PageProvider->getInheritedConfiguration(), which creates inheritance tree (array, comtaining all parents of PageL4). And then in foreach loop we get into a huge recursion:
1. Tree is iterated from root to leaf (Home->PageL1->PageL2->PageL3)
2. On each cycle ```$provider->getForm($branch);``` is called, which at defined point calls itself getInheritedPropertyValueByDottedPath() for each property, that current branch in cycle contains.
3. Then ```$provider->getFlexFormValuesSingle($branch);``` is called, which again at defined point calls getInheritedConfiguration() for each property.
4. Going deeper in a tree (iteration, mentioned at item 1) increases getInheritedConfiguration() calls, becasue now they are called not only twice for each broperty of current branchm but also each parent's property is called again.

So, in a situation above we go like this:
1. Read settings.footerPid from PageL4
1.1. Read all settings from Home and try to get their inherited values (which do not exist, becasue it is root).
1.2. Read all settings from PageL1
1.2.1. Read settings.footerPid from PageL1
1.2.1.1. Read settings.footerPid from Home 
1.2.1.2. Read settings.headerPid from Home 
1.2.1.3. Read settings.menuPid from Home 
1.2.1.4. Read settings.whateverPid from Home 
1.2.2. Read settings.headerPid from PageL1
1.2.2.1. Read settings.footerPid from Home 
1.2.2.2. Read settings.headerPid from Home 
1.2.2.3. Read settings.menuPid from Home 
1.2.2.4. Read settings.whateverPid from Home 
1.2.3. Read settings.menuPid from PageL1
1.2.3.1. Read settings.footerPid from Home 
1.2.3.2. Read settings.headerPid from Home 
1.2.3.3. Read settings.menuPid from Home 
1.2.3.4. Read settings.whateverPid from Home 
1.2.4. Read settings.whateverPid from PageL1
1.2.4.1. Read settings.footerPid from Home 
1.2.4.2. Read settings.headerPid from Home 
1.2.4.3. Read settings.menuPid from Home 
1.2.4.4. Read settings.whateverPid from Home 
1.3. Read all settings from PageL2
1.3.1. Read settings.footerPid from PageL2
1.3.1.1. Read settings.footerPid from PageL1
1.3.1.1.1. Read settings.footerPid from Home
1.3.1.1.2. Read settings.headerPid from Home
1.3.1.1.3. Read settings.menuPid from Home
1.3.1.1.4. Read settings.whateverPid from Home
1.3.1.2. Read settings.headerPid from PageL1
1.3.1.2.1. Read settings.footerPid from Home
1.3.1.2.2. Read settings.headerPid from Home
1.3.1.2.3. Read settings.menuPid from Home
1.3.1.2.4. Read settings.whateverPid from Home
...

You can see the problem. Basically, you can calculate number of recursion calls of getInheritedPropertyValueByDottedPath() and all the heavy methods, which are bound to this, with a sum of geometric progression, where each memeber is calculated by formula: ```a<sub>n</sub>=a*r<sup>n-1</sup>```
where a = 8 (calls on a root level)
r = 8 (calls on each level itself)
n = your current root level
And a sum is: ```a*(1-r<sup>n</sup>)/(1-r)```
Therefore in case above it'll be: ```8*(1-8<sup>5</sup>)/(1-5) = 65534```
That's too much and this is close, to what Xdebug profiler gives to me.

Sure, if you have more settings (bcs of using different templates) at some levels, then you can get even bigger numbers.

Possible fix: use at least runtime cache for getInheritedConfiguration() method to avoid useless method calls.

**Comment by bjo3rnf** [comment]
Fixed by #257
